43 implement playing blank cards

joshjwilli opened the pull request:
- Created a card prefab
- Cards are cast by dragging and letting go and player movement should be disabled while casting
- I adjusted the game screen for 9:16 or 1080x1920 (other devs might have to change their game screen resolution on their end for it to not look funky) this is so the formatting is closer to mobile phone screens.
- CardHandler.cs handles everything the card object needs to do while CardManager.cs handles the database of card abilities and handles communicating with PlayerController.cs
- The left card heals for 20, the middle card causes the player to take 20 damage, and the right card instantly kills the closest enemy but leaves the player at 1 hp.

joshjwilli committed:
Implemented all but tap and drag casting, adjusted UI for 9:16

- Readjusted the UI completely for 9:16 aspect ratio, aka average mobile screen to properly fit the cards
- Three cards are now implemented in the levels, click to cast with cooldowns. - The left card heals for 20, the middle card causes the player to take damage for 20, and the right card instantly kills the closest enemy and leaves the player at 1 health
- Player movement is disabled during casting, but casting is currently instant on press so no testing effect

joshjwilli committed:
Implemented Drag Casting

adenprince commented on Assets/Scripts/Game Controllers/CardManager.cs:
```diff
@@ -0,0 +1,47 @@
+    {
+        switch (card)
+        {
+            case "LeftCard":
```
Instead of using a card name and `CardManager` to decide what functionality to use when a card is played, `CardHandler` could be an abstract class with card subclasses that each have their own implementation of an abstract `TryPlayCard` method.
The Card Handler component on the Card prefab cannot be reused with that implementation, but I think doing that is simpler overall.

joshjwilli replied:
The current implementation is one abstract Card script with a script per card that inherits it. Is there a way to have it all in one script?

adenprince replied:
It could all be in one script by calling a method based on the value of a name variable or using an enum to distinguish card types. I think it's easier to maintain the code and add new cards when card behaviors are in separate scripts.

adenprince commented on Assets/Scripts/Game Controllers/CardManager.cs:
```diff
@@ -0,0 +1,47 @@
+    [SerializeField]
+    private PlayerController player;
+
+    public bool PlayCard(string card)
```
This could be renamed to `TryPlayCard` because it returns a boolean indicating success, but it doesn't matter much.

adenprince commented on Assets/Scripts/Object Controllers/CardHandler.cs:
```diff
@@ -0,0 +1,60 @@
+        CardIsPlayable();
+    }
+
+    private void CardIsPlayable()
```
I think `MakeCardPlayable` and `StartCardCooldown` would be better because those names indicate operations more clearly than the current names.

adenprince commented on Assets/Scenes/Levels/Infinite.unity:
Card Hand should be anchored to the bottom of the screen so that the cards are still positioned properly in different portrait resolutions. Top Bar should be anchored to the top of the screen for the same reason.
Card Hand and Top Bar should also be prefabs to prevent needing to change those GameObjects across multiple scenes. (Canvas (UI) and Score (TMP) should also be prefabs, but that's unrelated to your changes. I can make those prefabs if you don't want to on your branch. Making just Canvas (UI) a prefab instead of all of the UI elements would also work, but I prefer to modify UI elements separately from a canvas prefab in case those elements need to be reused elsewhere or moved.)

joshjwilli replied:
Having prefabs of existing prefabs children won't cause issues? I.e. modifying the child won't update the parent's version or vice versa?

adenprince replied:
It shouldn't cause issues. If you modify the child prefab, the parent prefab's instance of that child prefab will be updated. If you modify an instance of the child prefab in the parent prefab, then those changes can override what's in the child prefab.

adenprince commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -30,20 +31,23 @@ public class PlayerController : MonoBehaviour
 
     public int GetCurrentHealth() => currentHealth;
 
+    public void SetCanPlayerMove(bool boo) { canPlayerMove = boo; }
```
`boo` -> `canPlayerMove` (or something like that).

adenprince commented on Assets/Scripts/PlayerController.cs:
```diff
@@ -30,20 +31,23 @@ public class PlayerController : MonoBehaviour
         if (PauseController.Instance.GamePaused) {  return;  }
 
-        if (Input.touchCount > 0)
+        if (canPlayerMove && Input.touchCount > 0)
```
The player can still move while playing a card if they tap a destination position before playing a card. I think the player should stop moving immediately when a card starts being played, but whether to do that is subjective. To implement that behavior, `targetPosition` could be set to `rb2D.position` in `SetCanPlayerMove` when `canPlayerMove` is set to false.

adenprince commented on Assets/Scripts/Game Controllers/CardManager.cs:
```diff
@@ -0,0 +1,47 @@
+        if (enem != null)
+        {
+            enem.GetComponent<EnemyController>().Die();
+            player.TakeDamage(player.GetCurrentHealth() - 1);
```
It's clearer to add a method named `SetHealth` (or something like that) to `PlayerController.cs` and use that method here: `player.SetHealth(1);`.

adenprince reviewed (changes requested):
I left some comments about changes. I would prefer to see something similar to the structure I described for the card implementation, but I'm okay with the current card implementation. Also make sure to merge `premain` into your branch so that your branch has recent changes that are on `premain`.

joshjwilli committed:
Merge branch 'premain' into 43-implement-playing-blank-cards

joshjwilli committed:
Implemented Request Changes + Consistency Changes + Bug Fix

- Card system is now abstract class based
- Made each test card have it's own unique color patterns (Green for heal, red for self-damage, blue to instantly kill closest enemy)
- Blue card was also balanced considering health decay, being set to 1 health is basically instant death so it's now 10
- Threw the new enemy sprites on the enemy in levels (these will need to be properly implemented later just thought it'd be nice to look at something other than a blue box)
- Reordered and renamed some stuff for the sake of consistency (Controls a game object = Controller, Manages a game system = Manager, Handles a specific thing = Handler)
- Fixed Score Transition Speed either being slow as hell or instant.

adenprince commented on Assets/Scripts/Game Object/Card/Card.cs:
```diff
@@ -0,0 +1,64 @@
+    protected Image cardImage;
+    protected Button cardButton;
+    protected Vector2 cardOriginPos;
+    protected RectTransform rectTransform;
```
Some of these variables can be made private, but it doesn't matter that much. `TryPlayCard` can also be made protected.

adenprince commented on Assets/Scripts/Game Object/PlayerController.cs:
```diff
@@ -34,6 +35,49 @@ public class PlayerController : MonoBehaviour
+
+    public void TakeDamage(int damage)
+    {
+        SetHealth(currentHealth -= damage);
```
Unrelated to your changes: `-=` -> `-` here and `+=` -> `+` in the Heal method.

adenprince commented on Assets/Scripts/Game System/LevelEndTriggerManager.cs:
```diff
@@ -1,7 +1,7 @@
 using UnityEngine.SceneManagement;
 
-public class LevelEndTriggerController : MonoBehaviour
+public class LevelEndTriggerManager : MonoBehaviour
```
I think for class naming:
- Controller should be used for scripts that control the behavior of a single entity.
- Manager should be used for scripts that control multiple entities.
- Handler should be used for a class that mainly contains an event handler method or methods.

By entity, I mean an individual thing in the game that can consist of multiple GameObjects.

adenprince commented:
I don't think "Game Object" should be used for the name of a folder in the Scripts folder because any use of a script that contains a `MonoBehaviour` subclass is going to use a GameObject. I'm not sure what else to name the folder, though. I think the folder structure in the Scripts folder should be more similar to the folder structure in the Prefabs folder.

adenprince commented on Assets/Scripts/Game Object/Card/LeftCard.cs:
```diff
@@ -0,0 +1,8 @@
+public class LeftCard : Card
```
I think it would be good for the Card subclass names to relate to their functionality.

adenprince commented on Assets/Scripts/Handler/Button/PauseHandler.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    [SerializeField] private GameObject pauseMenuUI;
+
+    public void PauseButton_OnClick()
```
I think that `Button_OnClick` should be omitted because the name of this class implies that it's not only for a button. (Other handler classes are similar.)

joshjwilli replied:
I agree that the naming should be adjusted, but should the handlers still be considered handlers or do you want them to be called something else?

adenprince replied:
I think it's fine for these scripts to still be named handlers.

adenprince reviewed (changes requested):
Looks good to me, thank you. I made some comments that you should take a look at. In general, I think cleanup that affects a lot of files should be done separately from implementing a feature to simplify code review and reduce the time that pull requests are open.

joshjwilli committed:
Implemented Requested Changes

Unsure how Obstacle Random Sprite should be named/organized

adenprince commented on Assets/Scripts/Handler/PauseHandler.cs:
```diff
@@ -1,10 +1,12 @@
-}
+using UnityEngine;
+
+public class PauseHandler : MonoBehaviour
```
This is more accurate:
`PauseHandler` -> `TogglePauseMenuHandler`
`Pause()` -> `TogglePauseMenu()`

joshjwilli replied:
Ill do that right now

adenprince reviewed (changes requested):
Your changes look good to me. I think the script organization is good. You need to update the Button components in prefabs and scenes to use the updated handler method names. Buttons aren't working because they reference methods that no longer exist.

joshjwilli committed:
Buttons Fixed

joshjwilli committed:
Changed Script Name

adenprince reviewed (approved):
Looks good to me. Thank you for making the requested changes.
